Fix is_iso_date pattern and add tests

ferdnyc opened the pull request:
Fixes #14

ferdnyc committed:
Fix is_iso_date pattern

Signed-off-by: FeRD (Frank Dana) <ferdnyc@gmail.com>

pombredanne commented:
@ferdnyc Thanks! good catch... it also means we should have a few tests for this pattern. Do you mind to add some?

ferdnyc committed:
Add tests for is_iso_date

Signed-off-by: FeRD (Frank Dana) <ferdnyc@gmail.com>

ferdnyc commented:
@pombredanne 

Done. (Along with a commit to add `setuptools` to the `testing` extra's dependencies, since it's needed to run `pytest`.)

I've confirmed that they pass on this branch, but fail when cherry-picked into `main` without the fix in my 2dec77f.

ferdnyc commented on tests/test_saneyaml.py:
```diff
@@ -137,6 +137,19 @@ def test_load_ignore_aliases(self):
+        assert not saneyaml.is_iso_date('1994-01-1')
+        assert not saneyaml.is_iso_date('1994-1-1')
+        assert not saneyaml.is_iso_date('193')
+
```
Hmm, I wonder if I should add this to the end of the tests?
```suggestion
        # Hey, nothing's perfect
        assert saneyaml.is_iso_date('2003-02-29')
        assert saneyaml.is_iso_date('2004-01-35')

```

pombredanne replied:
Yes, this is weird, but it is best to capture also this weirdness in tests.

pombredanne commented on setup.cfg:
```diff
@@ -55,6 +55,7 @@ where = src
 
 [options.extras_require]
 testing =
+    setuptools >= 50
```
Why do you need this? We do not have any hard dependency in tests on setuptools, do we?

And FWIW, we typically use this base repo and we merge it for the boilerplate files: https://github.com/nexB/skeleton

In all cases we are progressively moving away from setuptools to use a new build tool called flot https://github.com/nexB/flot that is grown from flit.

ferdnyc replied:
Apparently, yes. This fails with a missing `setuptools`:

```bash
$ python3 -m venv venv3.12
$ . ./venv3.12/bin/activate
$ pip install -e '.[testing]'
$ pytest
```

The build system depending on `setuptools` doesn't install it when _not_ building the package, is the problem.

ferdnyc replied:
The `pytest` output without `setuptools`, specifically, is:

```text
============================= test session starts ==============================
platform linux -- Python 3.12.3, pytest-8.2.2, pluggy-1.5.0
rootdir: /var/tmp/saneyaml
configfile: pyproject.toml
plugins: xdist-3.6.1
collected 24 items / 1 error                                                   

==================================== ERRORS ====================================
__________________________ ERROR collecting setup.py ___________________________
ImportError while importing test module '/var/tmp/saneyaml/setup.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/lib64/python3.12/importlib/__init__.py:90: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
setup.py:3: in <module>
    import setuptools
E   ModuleNotFoundError: No module named 'setuptools'
=========================== short test summary info ============================
ERROR setup.py
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
=============================== 1 error in 0.35s ===============================
```

ferdnyc replied:
Arguably `pytest` shouldn't be picking up `setup.py`, so an exclusion in the `pytest` config in `setup.cfg` or `pyproject.yaml` might be an alternate way to go.

ferdnyc replied:
(Looks like it's picking it up because of this:)

https://github.com/nexB/saneyaml/blob/40e5fa7c0b6e0012452053839184e5cd29802063/pyproject.toml#L10
https://github.com/nexB/saneyaml/blob/40e5fa7c0b6e0012452053839184e5cd29802063/pyproject.toml#L43

ferdnyc replied:
Anyway, I'll drop that commit, since it only affects local testing.

pombredanne replied:
An exclusion in setup.cfg looks best indeed.

pombredanne committed:
Add weird is_iso_date tests #14

These are non-valid ISO dates but we do not validate fully
just a surface validation using a regex.

This captures the weird cases we accept.

Reference: https://github.com/nexB/saneyaml/issues/14
Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

Co-authored-by: Frank Dana <ferdnyc@gmail.com>
Signed-off-by: FeRD (Frank Dana) <ferdnyc@gmail.com>

pombredanne reviewed (changes requested):
Thanks the tests! ... just a tiny nit wrt. setuptools

pombredanne reviewed (approved):
Thanks! Merging
